src/control/build/interop.go

```diff
@@ -0,0 +1,139 @@
+//
+// (C) Copyright 2022 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package build
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	// MaxMinorDelta is the default maximum minor version delta
+	// allowed between two components.
+	MaxMinorDelta = 2
+)
+
+// Component is a component of the system.
+type Component string
+
+// Matches returns true if the component is equal to other
+// or one of the components is a wildcard.
+func (c Component) Matches(other Component) bool {
+	return c == other || c == ComponentAny || other == ComponentAny
+}
+
+func (c Component) String() string {
+	return string(c)
+}
+
+var (
+	// ComponentAny is a wildcard component.
+	ComponentAny = Component("")
+	// ComponentServer represents the control plane server.
+	ComponentServer = Component("server")
+	// ComponentAdmin represents the Control API client.
+	ComponentAdmin = Component("admin")
+	// ComponentAgent represents the compute node agent.
+	ComponentAgent = Component("agent")
+)
+
+// NewVersionedComponent creates a new VersionedComponent.
+func NewVersionedComponent(comp Component, version string) (*VersionedComponent, error) {
+	v, err := NewVersion(version)
+	if err != nil {
+		return nil, err
+	}
+
+	switch comp {
+	case ComponentServer, ComponentAdmin, ComponentAgent, ComponentAny:
+		return &VersionedComponent{
+			Component: comp,
+			Version:   *v,
+		}, nil
+	default:
+		return nil, errors.Errorf("invalid component %q", comp)
+	}
+}
+
+// VersionedComponent is a component with a version.
+type VersionedComponent struct {
+	Version   Version
+	Component Component
+}
+
+func (vc *VersionedComponent) String() string {
+	return fmt.Sprintf("%s:%s", vc.Component, vc.Version)
+}
+
+// InteropRule is a rule for checking compatibility between two
+// versioned components.
+type InteropRule struct {
+	Self          Component
+	Other         Component
```

Comment:
Unclear to me what purpose these two members serve. Is the idea to have different rules for agent to server, server to agent, etc?

Reply:
Yes, exactly. At the moment we're not using them because the interop rules are very simple and only version-based. In a previous iteration I had more complex rules for checking server<->server and agent<->server compatibility, but those aren't necessary right now.